src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,856 @@
+  }
+  switch (expectedType.kind) {
+    case 'primitive':
+    case 'variable':
+      return (actualType as Primitive).name !== (expectedType as Primitive).name
+    case 'function':
+      if (actualType.kind !== 'function') {
+        return true
+      }
+      // Check parameter types
+      const actualParamTypes = actualType.parameterTypes
+      const expectedParamTypes = expectedType.parameterTypes
+      if (actualParamTypes.length !== expectedParamTypes.length) {
+        return true
+      }
+      for (let i = 0; i < actualType.parameterTypes.length; i++) {
+        // Note that actual and expected types are swapped here
+        // as expected type should be subset of actual type for function arguments
+        if (hasTypeMismatchErrors(expectedParamTypes[i], actualParamTypes[i])) {
+          return true
+        }
+      }
+      // Check return type
+      return hasTypeMismatchErrors(actualType.returnType, expectedType.returnType)
+    case 'union':
+      const expectedSet = new Set(expectedType.types.map(formatTypeString))
+      // If not union type, expected set should contain actual type
+      if (actualType.kind !== 'union') {
+        return !expectedSet.has(formatTypeString(actualType))
+      }
+      // If both are union types, actual set should be a subset of expected set
+      const actualSet = new Set(actualType.types.map(formatTypeString))
+      for (const elem of actualSet) {
+        if (!expectedSet.has(elem)) {
+          return true
+        }
+      }
+      return false
+    default:
+      return true
+  }
+}
+
+/**
+ * Checks the types of the arguments of the given call expression.
+ * If number of arguments is different, add InvalidNumberOfArguments error to context and terminates early.
+ * Else, checks each argument against its expected type.
+ */
+function checkArgTypes(
+  node: es.CallExpression,
+  expectedTypes: Primitive[],
+  context: Context,
+  env: TypeEnvironment
+) {
+  const args = node.arguments
+  if (args.length !== expectedTypes.length) {
+    context.errors.push(new InvalidNumberOfArguments(node, expectedTypes.length, args.length))
+    return
+  }
+  for (let i = 0; i < expectedTypes.length; i++) {
+    const expectedType = expectedTypes[i]
+    const node = args[i]
+    const actualType = typeCheckAndReturnType(node, context, env)
+    checkForTypeMismatch(node, actualType, expectedType, context)
+  }
+}
+
+/**
+ * Converts type annotation node to its corresponding type representation in Source.
+ * If no type annotation exists, returns the "any" primitive type.
+ */
+function getAnnotatedType(
+  annotationNode: AnnotationTypeNode | TypeAliasDeclarationNode | undefined,
+  context: Context,
+  env: TypeEnvironment
+): Type {
+  if (!annotationNode) {
+    return tAny
+  }
+  const annotatedTypeNode = annotationNode.typeAnnotation
+  switch (annotatedTypeNode.type) {
+    case TSNodeType.TSFunctionType:
+      const fnTypeNode = annotatedTypeNode as FunctionTypeNode
```

Comment:
The typecasting here is unnecessary if we make use of TypeScript's [discriminated unions](https://www.typescriptlang.org/docs/handbook/typescript-in-5-minutes-func.html#discriminated-unions).

```ts
@@ -373,39 +373,37 @@ export type TypeAnnotation = {
   returnType?: AnnotationTypeNode
 }
 
-export interface BaseTypeNode extends es.BaseNode {
-  type: TSNodeType
-}
+export type BaseTypeNode = AnnotationTypeNode | FunctionTypeNode | UnionTypeNode | TypeAliasDeclarationNode | AsExpressionNode | TypeReferenceNode
 
-export interface AnnotationTypeNode extends BaseTypeNode {
+export interface AnnotationTypeNode extends es.BaseNode {
   type: TSNodeType.TSAnnotationType
   typeAnnotation: BaseTypeNode | FunctionTypeNode | UnionTypeNode | TypeReferenceNode
 }
 
-export interface FunctionTypeNode extends BaseTypeNode {
+export interface FunctionTypeNode extends es.BaseNode {
   type: TSNodeType.TSFunctionType
   parameters: NodeWithTypeAnnotation<es.Identifier>[]
   typeAnnotation: AnnotationTypeNode
 }
 
-export interface UnionTypeNode extends BaseTypeNode {
+export interface UnionTypeNode extends es.BaseNode {
   type: TSNodeType.TSUnionType
   types: BaseTypeNode[]
 }
 
-export interface TypeAliasDeclarationNode extends BaseTypeNode {
+export interface TypeAliasDeclarationNode extends es.BaseNode {
   type: TSNodeType.TSTypeAliasDeclaration
   id: es.Identifier
   typeAnnotation: BaseTypeNode
 }
 
-export interface AsExpressionNode extends BaseTypeNode {
+export interface AsExpressionNode extends es.BaseNode {
   type: TSNodeType.TSAsExpression
   expression: es.Node
   typeAnnotation: BaseTypeNode
 }
 
-export interface TypeReferenceNode extends BaseTypeNode {
+export interface TypeReferenceNode extends es.BaseNode {
   type: TSNodeType.TSTypeReference
   typeName: es.Identifier
 }
```